baseband/guppi/header.py

```diff
@@ -0,0 +1,379 @@
+
+Implements a GUPPIHeader class that reads & writes FITS-like headers from file.
+"""
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+import numpy as np
+import astropy.units as u
+from astropy.io import fits
+from astropy.time import Time
+
+
+__all__ = ['GUPPIHeader']
+
+
+class GUPPIHeader(fits.Header):
+    """GUPPI baseband file format header.
+
+    Parameters
+    ----------
+    *args : str or iterable
+        If a string, parsed as a GUPPI header from a file, otherwise
+        as for the OrderedDict baseclass.
+    verify : bool, optional
+        Whether to do minimal verification that the header is consistent with
+        the GUPPI standard.  Default: `True`.
+    mutable : bool, optional
+        Whether to allow the header to be changed after initialisation.
+        Default: `True`.
+    **kwargs
+        Any further header keywords to be set.
+
+    Notes
+    -----
+    Like `~astropy.io.fits.Header`, the initialiser does not accept keyword
+    arguments to populate an array - instead, one must pass an iterable. In
+    order to ensure keywords are kept in the right order, one should pass on
+    values as a tuple, not as a dict. E.g., to copy a header, one should not
+    do ``GUPPIHeader({key: header[key] for key in header})``, but rather::
+
+        GUPPIHeader(((key, header[key]) for key in header))
+
+    or, to also keep the comments::
+
+        GUPPIHeader(((key, (header[key], header.comments[key]))
+                      for key in header))
+    """
+
+    _properties = ('payload_nbytes', 'frame_nbytes', 'bps', 'nchan', 'npol',
+                   'sample_shape', 'sample_rate', 'sideband', 'overlap',
+                   'samples_per_frame', 'offset', 'start_time',
+                   'time')
+    """Properties accessible/usable in initialisation for all headers."""
+
+    _defaults = [('BACKEND', 'GUPPI'),
+                 ('BLOCSIZE', 0),
+                 ('STT_OFFS', 0),
+                 ('PKTIDX', 0),
+                 ('OVERLAP', 0),
+                 ('SRC_NAME', 'unset'),
+                 ('TELESCOP', 'unset'),
+                 ('PKTFMT', '1SFA'),
+                 ('NBITS', 8),
+                 ('NPOL', 1),
+                 ('OBSNCHAN', 1)]
+
+    def __init__(self, *args, **kwargs):
+        verify = kwargs.pop('verify', True)
+        mutable = kwargs.pop('mutable', True)
+        # Comments handled by fits.Header__init__().
+        super(GUPPIHeader, self).__init__(*args, **kwargs)
+        self.mutable = mutable
+        if verify:
+            self.verify()
+
+    def verify(self):
+        """Basic check of integrity."""
+        assert all(key in self for key in ('BLOCSIZE',
+                                           'PKTIDX'))
+        assert self['PKTFMT'] == '1SFA'
+
+    def copy(self):
+        """Create a mutable and independent copy of the header."""
+        newfitsheader = super(GUPPIHeader, self).copy()
```

Comment:
Weird, logically, the `super` call should return the correct class already. A bug in `io.fits`... see https://github.com/astropy/astropy/issues/7379